akka-stream/src/main/scala/akka/stream/impl/ActorRefSource.scala

```diff
@@ -0,0 +1,147 @@
+          failStage(failureMatcher(m))
+        case (_, m) if completionMatcher.isDefinedAt(m) ⇒
+          isCompleting = true
+          pump()
+        case (_, m: T) ⇒ // todo: requires classTag (breaks api)
+          if (isCompleting) {
+            log.warning(
+              "Dropping element because Status.Success received already, only draining already buffered elements: [{}] (pending: [{}])",
+              m,
+              buffer.used)
+          } else if (buffer == null) { // todo: remove - for backwards compatibility only!
+            if (isAvailable(out)) push(out, m)
+            else log.debug("Dropping element because there is no downstream demand: [{}]", m)
+          } else if (!buffer.isFull) {
+            buffer.enqueue(m)
+            pump()
+          } else
+            overflowStrategy match {
+              case s: DropHead ⇒
+                log.log(
+                  s.logLevel,
+                  "Dropping the head element because buffer is full and overflowStrategy is: [DropHead]")
+                buffer.dropHead()
+                buffer.enqueue(m)
+                pump()
+              case s: DropTail ⇒
+                log.log(
+                  s.logLevel,
+                  "Dropping the tail element because buffer is full and overflowStrategy is: [DropTail]")
+                buffer.dropTail()
+                buffer.enqueue(m)
+                pump()
+              case s: DropBuffer ⇒
+                log.log(
+                  s.logLevel,
+                  "Dropping all the buffered elements because buffer is full and overflowStrategy is: [DropBuffer]")
+                buffer.clear()
+                buffer.enqueue(m)
+                pump()
+              case s: DropNew ⇒
+                log.log(
+                  s.logLevel,
+                  "Dropping the new element because buffer is full and overflowStrategy is: [DropNew]")
+              case s: Fail ⇒
+                log.log(s.logLevel, "Failing because buffer is full and overflowStrategy is: [Fail]")
+                val bufferOverflowException =
+                  BufferOverflowException(s"Buffer overflow (max capacity was: $maxBuffer)!")
+                failStage(bufferOverflowException)
+              case _: Backpressure ⇒
+              // there is a precondition check in Source.actorRefSource factory method to not allow backpressure as strategy
+            }
+
+        case (_, m) ⇒
+          log.warning("Dropping unexpected element: [{}]", m)
+      }.ref
+
+      @tailrec
+      private def pump(): Unit = {
+        if (isAvailable(out) && buffer.nonEmpty) {
+          val msg = buffer.dequeue()
+          push(out, msg)
+          pump()
+        } else if (isCompleting && buffer.isEmpty) {
+          completeStage()
+        }
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          pump()
+        }
+      })
+    }
+
+    (stage, stage.ref)
+  }
```

Comment:
Not sure how good the test coverage was, if there was _any_ for the old source actor - a quick search doesn't show any, I think this really needs some test coverage:

* happy path
* poison pill
* downstream cancels
* abrupt termination
* overflowing the buffer with the various strategies
* sending elements after completion matcher has said to complete

Reply:
Actually there are quite some tests for the old source actor (used them to build in all the backwards compatibility hacks like poison pill, buffer sizes etc).

But I can see to extend them a little.

Reply:
Where are they located?

Reply:
`akka.stream.scaladsl.ActorRefSourceSpec` in `actor-stream-tests`

Reply:
Obviously. 🤦‍♂️